File: app/src/main/java/org/caojun/salmagundi/string/utils/PinyinUtils.java
package org.caojun.salmagundi.string.utils;

import android.text.TextUtils;

import com.socks.library.KLog;

import net.sourceforge.pinyin4j.PinyinHelper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by CaoJun on 2017/8/18.
 */

public class PinyinUtils {

    /**
     * 只判断首字拼音多音字
     * @param text
     * @return
     */
    public static String[] toPinyin1st(String text) {
        if (TextUtils.isEmpty(text)) {
            return null;
        }
        char c = text.charAt(0);
        String[] py = PinyinHelper.toHanyuPinyinStringArray(c);
        String[] pinyin;
        if (py == null) {
            //非汉字
            pinyin = new String[1];
            pinyin[0] = text;
        } else {
            pinyin = new String[py.length];
            for (int i = 0;i < py.length;i ++) {
                pinyin[i] = py[i] + text;
            }
        }
        return pinyin;
    }

    /**
     * 列出所有拼音的排列组合
     * @param text
     * @return
     */
    public static String[] toPinyin(String text) {
        if (TextUtils.isEmpty(text)) {
            return null;
        }
        String[][] pinyin = new String[text.length()][];
        for (int i = 0;i < text.length();i ++) {
            char c = text.charAt(i);
            String[] py = PinyinHelper.toHanyuPinyinStringArray(c);
            if (py == null) {
                //非汉字
                pinyin[i] = new String[1];
                pinyin[i][0] = String.valueOf(c);
            } else {
                pinyin[i] = py;
            }
        }
        return doCombination(pinyin);
    }

    /**
     * 两个数组做排列组合
     * @param a
     * @param b
     * @return
     */
    private static String[] doCombination(String[] a, String[] b) {
        List<String> list = new ArrayList<>();
        for (int i = 0;i < a.length;i ++) {
            for (int j = 0;j < b.length;j ++) {
                list.add(a[i] + b[j]);
            }
        }
        String[] strings = new String[list.size()];
        list.toArray(strings);
        return strings;
    }

    private static String[] doCombination(String[][] data) {
        String[] strings;
        if (data.length < 2) {
            return data[0];
        } else {
            strings = doCombination(data[0], data[1]);
            if (data.length == 2) {
                return strings;
            }
            for (int i = 2;i < data.length;i ++) {
                strings = doCombination(strings, data[i]);
            }
            return strings;
        }
    }
}

File: PinyinLibrary/src/main/java/org/caojun/pinyin/PinyinUtils.kt
package org.caojun.pinyin

import android.text.TextUtils
import net.sourceforge.pinyin4j.PinyinHelper
import net.sourceforge.pinyin4j.format.HanyuPinyinOutputFormat
import net.sourceforge.pinyin4j.format.HanyuPinyinToneType
import net.sourceforge.pinyin4j.format.HanyuPinyinVCharType

object PinyinUtils {

    /**
     * 汉字转拼音
     *
     * @param c
     * @return
     */
    fun toPinyinStringArray(c: Char): Array<String>? {
        val format = HanyuPinyinOutputFormat()
        format.toneType = HanyuPinyinToneType.WITHOUT_TONE
        format.vCharType = HanyuPinyinVCharType.WITH_U_UNICODE
        try {
            return PinyinHelper.toHanyuPinyinStringArray(c, format)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return null
    }

    /**
     * 汉字转拼音
     *
     * @param c
     * @param first 只返回第一个
     * @return
     */
    fun toPinyinString(c: Char, first: Boolean): String {
        val pinyin = toPinyinStringArray(c) ?: return c.toString()
        if (pinyin.isEmpty()) {
            return c.toString()
        }
        if (pinyin.size == 1) return pinyin[0]
        val sb = StringBuffer()
        sb.append(pinyin[0])
        if (first) return sb.toString()
        sb.append("(")
        for (i in 1 until pinyin.size) {
                sb.append(pinyin[i])
                if (i < pinyin.size - 1) sb.append(",")
            }
        sb.append(")")
        return sb.toString()
    }

    /**
     * 汉字转拼音
     *
     * @param text
     * @return
     */
    fun toPinyinString(text: String): String {
        if (TextUtils.isEmpty(text)) return text
        val sb = StringBuffer()
        for (i in 0 until text.length) {
            val c = text[i]
            val pinyin = toPinyinString(c, true)
            if (TextUtils.isEmpty(pinyin)) {
                sb.append(c)
            } else {
                sb.append(pinyin)
            }
            if (i < text.length - 1) {
                sb.append(" ")
            }
        }
        return sb.toString()
    }

    /**
     * 汉字转拼音首字母
     *
     * @param text
     * @return
     */
    fun toPinyinInitials(text: String): String {
        if (TextUtils.isEmpty(text)) return text
        val sb = StringBuffer()
        for (i in 0 until text.length) {
            val c = text[i]
            val pinyin = toPinyinString(c, true)
            if (TextUtils.isEmpty(pinyin)) {
                sb.append(c)
            } else {
                sb.append(pinyin[0])
            }
        }
        return sb.toString()
    }
}